Classes/common/HTTP/CDTURLSessionTask.m

```diff
@@ -108,36 +97,76 @@ - (NSURLSessionTaskState)state
 
 #pragma mark Helpers
 
+- (NSURLSessionDataTask *)makeRequest
+{
+    __block CDTHTTPInterceptorContext *ctx =
+        [[CDTHTTPInterceptorContext alloc] initWithRequest:[self.request mutableCopy]];
+
+    // We make sure all objects support `interceptRequestInContext:` during init.
+    for (NSObject<CDTHTTPInterceptor> *obj in self.requestInterceptors) {
+        ctx = [obj interceptRequestInContext:ctx];
+    }
+
+    __weak CDTURLSessionTask *weakSelf = self;
+    return [self.session
+        dataTaskWithRequest:ctx.request
+          completionHandler:^void(NSData *data, NSURLResponse *response, NSError *error) {
+            __strong CDTURLSessionTask *strongSelf = weakSelf;
+            if (strongSelf) {
+                ctx.response = response;
+                for (NSObject<CDTHTTPInterceptor> *obj in strongSelf.responseInterceptors) {
+                    ctx = [obj interceptResponseInContext:ctx];
+                }
+
+                if (ctx.shouldRetry && strongSelf.remainingRetries > 0) {
+                    // retry
+                    strongSelf.remainingRetries--;
+                    strongSelf.inProgressTask = [strongSelf makeRequest];
+                    [strongSelf.inProgressTask resume];
+                } else if (strongSelf.completionHandler) {
+                    data = [NSData dataWithData:data];
```

Comment:
I read the Documentation for URL session if a task is a NSURLSessionDownloadTask, the data backing the object, would be a file which is deleted when the completion block ends. At the time I initially thought this would prevent that, but actually I'm not sure. I'll take that assignment out